### PR TITLE
[Diagnostics] Improve the diagnostic for invalid optional pointer conversions for an autoclosure result type.

### PR DESCRIPTION
**Pull request by hborla**
Skip `ConstraintLocator::OptionalPayload` when checking if we're in an autoclosure context for an inout-to-pointer conversion. This lets us apply the `AllowAutoClosurePointerConversion` constraint fix when the pointer is optional.

rdar://problem/53532404
https://bugs.swift.org/browse/SR-11198

### CODE REVIEW

**Comment by hborla** [comment]
@swift-ci please smoke test

File: lib/Sema/ConstraintLocator.h [hunk 1]
```diff
@@ -715,6 +715,21 @@ class ConstraintLocatorBuilder {
     return (getSummaryFlags() & ConstraintLocator::IsFunctionConversion);
   }
 
+  bool isAutoclosureResult() const {
+    SmallVector<LocatorPathElt, 4> path;
+    getLocatorParts(path);
+
+    auto last = std::find_if(
+        path.rbegin(), path.rend(), [](LocatorPathElt &elt) -> bool {
+          return elt.getKind() != ConstraintLocator::OptionalPayload;
```
Comment (hamishknight) [inline lib/Sema/ConstraintLocator.h]: I think we'll also want to look through `ConstraintLocator::GenericArgument` as well in order to handle optional-to-optional conversions such as:

```swift
func foo(_ x: @autoclosure () -> UnsafePointer<Int>?) {}

var x: [Int]? = []
foo(x)
```

**Review by hamishknight — COMMENTED** [review]
Thanks for fixing this! Is it worth guarding the array-to-pointer and string-to-pointer cases with a language version check, given it's technically a source-breaking change? For example, the following are currently valid:

```swift
func foo(_ x: @autoclosure () -> UnsafePointer<Int8>?) {}

var str: String = ""
foo(str)

var arr: [Int8] = []
foo(arr)
```

Though I somewhat doubt many people are relying on this.

File: test/Constraints/invalid_implicit_conversions.swift [hunk 2]
```diff
@@ -23,6 +24,7 @@ func test(
   takesAutoclosure(mutRawPtr, &a) // expected-error {{cannot perform pointer conversion of value of type '[Int]' to autoclosure result type 'UnsafeMutableRawPointer'}}
   takesAutoclosure(mutPtr, &a) // expected-error {{cannot perform pointer conversion of value of type '[Int]' to autoclosure result type 'UnsafeMutablePointer<Int>'}}
   takesAutoclosure(ptr, &a) // expected-error {{cannot perform pointer conversion of value of type '[Int]' to autoclosure result type 'UnsafePointer<Int>'}}
+  takesAutoclosure(optPtr, &i) // expected-error {{cannot perform pointer conversion of value of type 'Int' to autoclosure result type 'UnsafePointer<Int>?'}}
```
Comment (xedin) [inline test/Constraints/invalid_implicit_conversions.swift]: This is effectively a source break even though it's in an invalid code which should have been rejected, I think we might want to document this in `CHANGELOG` or similar document...

File: lib/Sema/ConstraintLocator.h [hunk 3]
```diff
@@ -715,6 +715,21 @@ class ConstraintLocatorBuilder {
     return (getSummaryFlags() & ConstraintLocator::IsFunctionConversion);
   }
 
+  bool isAutoclosureResult() const {
```
Comment (xedin) [inline lib/Sema/ConstraintLocator.h]: I think it make since to rename this to isForAutoClosureResult() just like we have it for isForKeyPathComponent.

**Review by xedin — APPROVED** [review]

File: lib/Sema/ConstraintLocator.h (hunk 1 above)
Comment (xedin) [reply lib/Sema/ConstraintLocator.h]: This is going to be a source break as well just like `Int` to `UnsafePointer<Int>?` in one of the tests. We might need to get Core Team involved in this. /cc @DougGregor

Comment (xedin) [reply lib/Sema/ConstraintLocator.h]: @hborla See https://github.com/apple/swift/pull/26543 that's one of the places where we'd like to add a new check for autoclosure.

File: test/Constraints/invalid_implicit_conversions.swift (hunk 2 above)
Comment (hborla) [reply test/Constraints/invalid_implicit_conversions.swift]: This particular example doesn't compile currently with the error `'&' can only appear immediately in a call argument list`. But this change is still a source break with the examples above

Comment (xedin) [reply test/Constraints/invalid_implicit_conversions.swift]: I tried this simplified version:
```swift
func foo(_: @autoclosure () -> UnsafePointer<Int>?) {}
var i: Int = 0
foo(&i)
```

Comment (xedin) [reply test/Constraints/invalid_implicit_conversions.swift]: Ok interesting, looks like it has been diagnosed in 4.2, 5.0 and 5.1 (one of the snapshots  I have) which means that it's only a master regression.

**Comment by hborla** [comment]
@swift-ci please smoke test

File: lib/Sema/ConstraintLocator.h (hunk 1 above)
Comment (DougGregor) [reply lib/Sema/ConstraintLocator.h]: The Core Team is okay with accepting this source break given that this code never should have compiled in the first place is can trigger undefined behavior. Thanks for your patience.

**Pull request merged**